### DISCUSSION TITLE
FR: `document formatting` capability for the `biome lsp-proxy`

### DISCUSSION THREAD

**Discussion by chrisgrieser** [biomejs/biome#87] [Ideas]
Not sure which issue type this belongs to, so opening a discussion here first.

While `biome` doubles as a CLI and as an LSP, only the linting/analyzing capabilities are accessible via CLI and LSP. Formatting capabilities are only accessible via CLI, since biome apparently does not register the document formatting lsp server capability. 

So for the editor integration of biome, this results in the somewhat awkward situation where I integrate linting via the lsp-client, and formatting via the formatting-client of my editor. (Which, in the case of nvim, require totally different plugins.)

---

relevant docs: https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_formatting

**Comment by ematipico** [comment]
Biome **does** support formatting via LSP

https://github.com/biomejs/biome/blob/dd9330fb4752b76483a6c2278c3aec5ccf66a802/crates/rome_lsp/src/server.rs#L148-L156

It's possible that your client is not able to register to this event because it doesn't support dynamic registration of the events

https://github.com/biomejs/biome/blob/dd9330fb4752b76483a6c2278c3aec5ccf66a802/crates/rome_lsp/src/session.rs#L358-L365

**Reply by chrisgrieser** [reply]
Thanks for the quick reply!

So, that's odd. Looking through the server configs of biome and tsserver, biome does not have `server_capabilities.documentFormattingProvider`, while tsserver does. 

Here the full server configs:
[tsserver-lsp-server-info.txt](https://github.com/biomejs/biome/files/12473966/tsserver-lsp-server-info.txt)
[biome-lsp-server-info.txt](https://github.com/biomejs/biome/files/12473967/biome-lsp-server-info.txt)

Though I am not familiar enough with lsp internals to know how dynamic registration plays a role here. As far as I can tell, most lsp servers I use have formatting capabilities without any issue? 🤔

**Reply by ematipico** [reply]
I'd suggest looking at third party LSP clients that support rome: https://github.com/neovim/nvim-lspconfig/blob/master/lua/lspconfig/server_configurations/rome.lua

**Reply by chrisgrieser** [reply]
well, before installing an additional lsp client, I might as well stick with using rome/biome's formatting via the cli and a formatting client 😅 

Wouldn't it be possible to have biome register the formatting capability in a non-dynamic way, which apparently most LSP servers do?

**Reply by ematipico** [reply]
> well, before installing an additional lsp client, I might as well stick with using rome/biome's formatting via the cli and a formatting client 😅

Sorry, what I meant to suggest was to **look at other clients** to see how they do setup biome LSP. 

Unfortunately, for us it's a must to dynamically register the capabilities because if the configuration changes, we must apply the new options.

**Reply by chrisgrieser** [reply]
> for us it's a must to dynamically register the capabilities because if the configuration changes, we must apply the new options.

Why though? I do not think that people change configs very often, especially during editing. Being aware of config changes during runtime seems at most like a nice-to-have for me, not really a "must"? I mean, again, it's the same with other LSPs, where the majority apparently does not dynamically register formatting capabilities. 

Like, not meaning to bother you (especially in this transition period from rome to biome), I'll just stick with the formatter client, but I am just curious why rome/biome is so fundamentally different from other LSPs in this regard.

**Reply by ematipico** [reply]
The original mission of the project is being a full fledged LSP, by providing not just formatting and linting, but also - in the future - code snippets, renaming capabilities, auto complete, code suggestions, etc. Very much like the typescript language server.

In order to do that, the LSP needs to have a client able to dynamically register capabilities.

You are right that the configuration might not change very much, but if a new user is setting up the linter - with more than 150 rules - do you think it's a good DX to restart the client at every change? I would expect to see my changes being applied instantly (for your example downgrading a rule to warning).

Also, the VSCode extension, which is directly maintained by us, has some options that require dynamic registration.

**Reply by musjj** [reply]
Formatting works for me using the default `lspconfig` setup. But I'm not sure how to sort imports. It seems that import sorting is done using `biome check`, but I can't figure out how to invoke it via LSP.

**Reply by chrisgrieser** [reply]
@musjj which nvim version are you on? [It seems the capability in question works in the nightly (0.10)](https://github.com/neovim/nvim-lspconfig/issues/2807)

**Reply by musjj** [reply]
I'm using `v0.10.0-dev-9ec0ecb`. Formatting works, but I can't get it to sort imports.

**Reply by chrisgrieser** [reply]
I see, yeah, so it (mostly) works in nvim 0.10, so this can be considered solved I guess. 

(The not-working import-sort is maybe worth a bug report?)

**Reply by musjj** [reply]
I still can't tell if it's a bug yet. In `biome`, import sorting is not a part of the formatter, unlike `prettier`, so the formatter understandably doesn't sort the imports. Not sure if this is possible via the LSP or if it's just not implemented yet. Tbh, I wish they'd just make import sorting a part of the formatter.

**Reply by Conaclos** [reply]
Import sorting is not part of the formatter because it can produce a code that behaves differently; Although this is rare. You can organize imports via the LSP thanks to the dedicated action `organize import`. On CLI, you must apply unsafe fixes using `biome check --apply-unsafe`. In the next release, the code fix will be flagged as safe. You will be able to execute it using ``biome check --apply`.

**Reply by musjj** [reply]
I tried to view the available code actions, but I can't find the `organize import` action. I know the biome LSP is functioning, because I can see code actions for other linting actions.

I also tried to move my cursor towards the top of the file (where the imports are), but I also still can't trigger that specific code action. I'm using `1.2.2`, maybe this feature is available in later versions?

**Reply by binarious** [reply]
Have you found a solution? And also: Does biome raise formatting violations as diagnostics for you? I only see linting issues. `vim.lsp.buf.format` works though.

Edit: Just doesn't seem to be implemented: https://github.com/biomejs/biome/discussions/809

**Comment by ematipico** [comment]
@chrisgrieser Do you still need this feature? I think we are going to drop the `requireConfiguration` option we have in the VSCode extension, this means we can drop the dynamic capability of the LSP.

**Reply by chrisgrieser** [reply]
yes, would still be nice! Though it's not high priority for me, since nvim 0.10 (still nightly) will support dynamic capabilities.

Nevertheless, if it is possible to drop the requirement, doing so might be useful for other editors that do not have complete LSP support, like nvim prior to 0.10?
